### PR TITLE
user profile side branch

### PR DESCRIPTION
**Pull request by kaydee0502**

### CODE REVIEW

**Review by priyamishra2903 — APPROVED** [review]

File: devsnest/Gemfile [hunk 1]
```diff
@@ -55,3 +56,6 @@ gem 'sidekiq-cron'
 gem 'httparty'
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
+
+gem 'mimemagic', github: 'mimemagicrb/mimemagic', ref: '01f92d86d15d85cfd0f20dabd025dcbd36a8a60f'
+gem 'rubocop', '~> 1.15', require: false
```
Comment (vedansh) [inline devsnest/Gemfile:61]: do we need it

File: devsnest/app/controllers/api/v1/college_controller.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/college_controller.rb:1]: I suppose we are not using this controller anymore

File: devsnest/app/models/college.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class College < ApplicationRecord
+
+  def self.create_college(college_name)
+    college = College.new
+    college.name = college_name
+    p college_name
+    college.save
+  end
+end
```
Comment (vedansh) [inline devsnest/app/models/college.rb:9]: add a new line

File: devsnest/app/models/college.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class College < ApplicationRecord
+
+  def self.create_college(college_name)
+    college = College.new
```
Comment (vedansh) [inline devsnest/app/models/college.rb:4]: do this in one line
`College.create(name: college_name)`

File: devsnest/app/models/content.rb [hunk 5]
```diff
@@ -5,4 +5,28 @@ class Content < ApplicationRecord
   enum difficulty: %i[easy medium hard]
   enum question_type: %i[class assignment]
   has_many :submission
+
+
+  def self.split_by_difficulty
+    questions = Content.all.pluck(:difficulty)
```
Comment (vedansh) [inline devsnest/app/models/content.rb:11]: `Content.where(data_type: "question").group(:difficulty).count`

We should try to do these operations via sql

File: devsnest/app/models/group.rb [hunk 6]
```diff
@@ -3,4 +3,13 @@
 class Group < ApplicationRecord
   # belongs_to :batch
   has_many :group_members
+
+  def self.update_group_name(user_id, group_name)
```
Comment (vedansh) [inline devsnest/app/models/group.rb:7]: Why do we need this

File: devsnest/app/models/submission.rb [hunk 7]
```diff
@@ -51,4 +51,25 @@ def self.recalculate_all_scores
       user.save
     end
   end
+
+  def self.count_solved(user_id)
+    content = Submission.where(user_id: user_id, status: 0).pluck(:content_id)
+    solves = Hash["easy" => 0,"medium" => 0,"hard" => 0]
```
Comment (vedansh) [inline devsnest/app/models/submission.rb:57]: Like above, do most of this with sql

File: devsnest/app/resources/api/v1/college_resource.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+module Api
+    module V1
+      class CollegeResource < JSONAPI::Resource
+        attributes :id,:name
+        
+      end
+    end
+  end
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/college_resource.rb:8]: new line

File: devsnest/app/controllers/api/v1/college_controller.rb (hunk 2 above)
Comment (vedansh) [reply devsnest/app/controllers/api/v1/college_controller.rb:1]: or we are only using it for index and not create

File: devsnest/db/migrate/20210520142304_add_column_to_user.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class AddColumnToUser < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :batch, :string
+    add_column :users, :grad_status, :string
+    add_column :users, :grad_specialization, :string
+    add_column :users, :grad_year, :integer
```
Comment (vedansh) [inline devsnest/db/migrate/20210520142304_add_column_to_user.rb:6]: can we merge all migration files that you created related to one source, like one migration file for all user related changes

File: devsnest/config/routes.rb [hunk 10]
```diff
@@ -17,6 +17,7 @@
       jsonapi_resources :submissions, only: %i[create]
       jsonapi_resources :groups, only: %i[show]
       jsonapi_resources :group_members, only: %i[index show]
+      jsonapi_resources :college, only: %i[index create]
     end
```
Comment (vedansh) [inline devsnest/config/routes.rb:21]: are we using `create`

File: devsnest/config/database.yml [hunk 11]
```diff
@@ -13,6 +13,7 @@ default: &default
   username: <%= ENV["MYSQL_USERNAME"]  %>
   password: <%= ENV["MYSQL_PASSWORD"] %>
   encoding: utf8
+  #socket: /var/run/mysqld/mysqld.sock
```
Comment (vedansh) [inline devsnest/config/database.yml:16]: remove

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 12]
```diff
@@ -3,23 +3,68 @@
...
+      end   
+
+      def total_by_difficulty
+        return nil if context[:user].nil?
+        Content.split_by_difficulty
+      end
+      
+      def activity
+        return nil if context[:user].nil?
+        dates = Hash.new
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/user_resource.rb:58]: Create a function in submission, get the data from there, do not write business logic here, secondly try to use as much as sql as possible

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 13]
```diff
@@ -3,23 +3,68 @@
...
+      def group_name
+        return nil if context[:user].nil?
+        member = GroupMember.where(user_id: context[:user].id).first 
+        groupId = member.present? ? member.group_id : nil
+        group =  Group.where(id: groupId).first
+        group.present? ? group.name : nil
+      end
+
+      def college_name
+        return nil if context[:user].nil?
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/user_resource.rb:40]:
```return nil if context[:user].nil? || context[:user].college_id.nil?    
     College.where(id: context[:user].college_id ).first.name || nil
```

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 14]
```diff
@@ -3,23 +3,68 @@
...
+      def total_by_difficulty
+        return nil if context[:user].nil?
+        Content.split_by_difficulty
+      end
+      
+      def activity
+        return nil if context[:user].nil?
+        dates = Hash.new
+        Submission.where(status:"done",user_id:context[:user].id).or(Submission.where(status:"doubt",user_id:context[:user].id)).all.each do |user|
+          if dates.key?(user.updated_at.to_date)
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/user_resource.rb:60]: `Submission.where(status:["done", "doubt"],  user_id:context[:user].id)`

File: devsnest/app/controllers/api/v1/college_controller.rb (hunk 2 above)
Comment (kaydee0502) [reply devsnest/app/controllers/api/v1/college_controller.rb:1]: Yeah, i created this controller to index all colleges using it's route

File: devsnest/app/models/college.rb (hunk 4 above)
Comment (kaydee0502) [reply devsnest/app/models/college.rb:4]: Redundant code : i am now creating college with JSON API Res, removing this snippet

File: devsnest/app/models/college.rb (hunk 3 above)
Comment (kaydee0502) [reply devsnest/app/models/college.rb:9]: Redundant code : i am now creating college with JSON API Res, removing this snippet

File: devsnest/app/models/content.rb (hunk 5 above)
Comment (kaydee0502) [reply devsnest/app/models/content.rb:11]: Done using this:

```ruby
  def self.split_by_difficulty
    questions = Content.group(:difficulty).count
  end
```

File: devsnest/app/models/group.rb (hunk 6 above)
Comment (kaydee0502) [reply devsnest/app/models/group.rb:7]: these are added by @priyamishra2903

File: devsnest/app/models/college.rb (hunk 4 above)
Comment (vedansh) [reply devsnest/app/models/college.rb:4]: aren't you using this function is user update

File: devsnest/app/models/submission.rb (hunk 7 above)
Comment (kaydee0502) [reply devsnest/app/models/submission.rb:57]: Done using:
```rb
 def self.count_solved(user_id)
    content = Submission.where(user_id: user_id, status: 0).pluck(:content_id)
    Content.where(id: content).group(:difficulty).count
  end  
```

File: devsnest/config/routes.rb (hunk 10 above)
Comment (kaydee0502) [reply devsnest/config/routes.rb:21]: Yes we are using it to let user from frontend to add their college if it's not listed in the dropdown.

File: devsnest/app/resources/api/v1/college_resource.rb (hunk 8 above)
Comment (kaydee0502) [reply devsnest/app/resources/api/v1/college_resource.rb:8]: done

File: devsnest/config/database.yml (hunk 11 above)
Comment (kaydee0502) [reply devsnest/config/database.yml:16]: done

File: devsnest/app/resources/api/v1/user_resource.rb (hunk 12 above)
Comment (kaydee0502) [reply devsnest/app/resources/api/v1/user_resource.rb:58]: done

File: devsnest/app/resources/api/v1/user_resource.rb (hunk 13 above)
Comment (kaydee0502) [reply devsnest/app/resources/api/v1/user_resource.rb:40]: done

File: devsnest/app/resources/api/v1/user_resource.rb (hunk 14 above)
Comment (kaydee0502) [reply devsnest/app/resources/api/v1/user_resource.rb:60]: done

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 15]
```diff
@@ -4,67 +4,63 @@ module Api
...
+          super - %i[password college_id]
         else
-          super - [:password, :email, :college_id]
+          super - %i[password email college_id]
         end
       end
 
       def self.updatable_fields(context)
-        super - [:score, :group_id, :group_name]
+        super - %i[score group_id group_name]
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/user_resource.rb:22]: discord_id, password, also check for all other fields,

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 16]
```diff
@@ -4,67 +4,63 @@ module Api
...
       end
 
       def group_name
         return nil if context[:user].nil?
-        member = GroupMember.where(user_id: context[:user].id).first 
-        groupId = member.present? ? member.group_id : nil
-        group =  Group.where(id: groupId).first
+
+        member = GroupMember.where(user_id: context[:user].id).first
+        group_id = member.present? ? member.group_id : nil
```
Comment (vedansh) [inline devsnest/app/resources/api/v1/user_resource.rb:36]: replace these 3 bottom lines with this 
`member.group.name`

File: devsnest/app/resources/api/v1/user_resource.rb (hunk 15 above)
Comment (kaydee0502) [reply devsnest/app/resources/api/v1/user_resource.rb]: done, what else fields do we need to include?

File: devsnest/app/resources/api/v1/user_resource.rb (hunk 16 above)
Comment (kaydee0502) [reply devsnest/app/resources/api/v1/user_resource.rb]: done

**Review by vedansh — APPROVED** [review]

**Pull request merged**